File: javascript/leetcode/advanced-algorithm/divide-and-qonquer/majority-element.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var majorityElement = function (nums) {
  if (nums.length === 0) return null;

  if (nums.length === 1) return nums[0];

  const half = Math.floor(nums.length / 2);
  const left = majorityElement(nums.slice(0, half));
  const right = majorityElement(nums.slice(half));

  const leftFilterOfNums = nums.filter((num) => num === left);
  const majorityIndex = leftFilterOfNums.length > half ? 1 : 0;
  return [right, left][majorityIndex];
};

console.log(majorityElement([2, 2, 1, 1, 1, 2, 2]));

File: python/leetcode/advenced-algorithm/divide-and-conquer/majority-element.py
from typing import List


class Solution:
    def majorityElement(self, nums: List[int]):
        if not nums:
            return
        length = len(nums)

        if length == 1:
            return nums[0]

        a = self.majorityElement(nums[: length // 2])
        b = self.majorityElement(nums[length // 2 :])

        print(nums)
        print(nums.count(a))
        return [b, a][nums.count(a) > length // 2]


print(Solution().majorityElement([2, 2, 1, 1, 1, 2, 2]))
